src/Simulation/Simulators/QuantumSimulator/StateDumper.cs

```diff
@@ -155,25 +323,163 @@ public virtual string Format(uint idx, double real, double img)
 
             }
 
+            public override string ToString()
```

Comment:
For the eventual IQ# PR, we may want to refactor its plain-text display encoder to use the new implementation here so that it can reuse `FormatCartesian` and `FormatPolar` instead of re-implementing. If we do that, it may be worth making `ToString` call a new overload of `ToString` that allows passing more options. That could likely be deferred until we have the matching IQ# PR closer to being ready to merge in.

Reply:
I assume you mean something like this:
```c#
            public string ToString(BasisStateLabelingConvention convention,   // Non-override. Parameterized.
                                   bool truncateSmallAmplitudes, 
                                   double truncationThreshold)
            {
                return string.Join('\n', 
                    SignificantAmplitudes(convention, truncateSmallAmplitudes, truncationThreshold)
                        .Select(
                            item =>
                            {
                                var (cmplx, basisLabel) = item;
                                var amplitude = (cmplx.Real * cmplx.Real) + (cmplx.Imaginary * cmplx.Imaginary);
                                var angle = System.Math.Atan2(cmplx.Imaginary, cmplx.Real);

                                return $"|{basisLabel}⟩\t{FormatCartesian(cmplx.Real, cmplx.Imaginary)}\t == \t" +
                                                    $"{FormatPolar(amplitude, angle)}";
                            }));
            }

            public override string ToString()   // An override of the `object.ToString()`.
            {
                return ToString(BasisStateLabelingConvention.LittleEndian, true, 0.8);   // Fixed values!!!
            }
```
Would you recommend to make this change now in this PR?

Reply:
As mentioned, I don't think it's urgent to include in this PR. At the same there's also no harm in doing so.

```csharp
var amplitude = (cmplx.Real * cmplx.Real) + (cmplx.Imaginary * cmplx.Imaginary);
var angle = System.Math.Atan2(cmplx.Imaginary, cmplx.Real);
```

As a side note, I believe that System.Numerics.Complex already has properties for that, such that the `Select` expression can be significantly simplified.

Reply:
```csharp
public override string ToString() => // An override of the `object.ToString()`.
    ToString(BasisStateLabelingConvention.LittleEndian, false, 0.0);   // Fixed values!!!
```

Setting truncation to `false` should reproduce previous behavior; the threshold is ignored in that case.

Reply:
> As a side note, I believe that System.Numerics.Complex already has properties for that, such that the Select expression can be significantly simplified.

Based on [System.Numerics Methods](https://docs.microsoft.com/en-us/dotnet/api/system.numerics.complex?view=net-5.0#methods), here is what I've come up with, not sure if it is absolutely right (already forgot a lot of Complex arithmetic):
```c#
var amplitude =  Abs(cmplx) * Abs(cmplx);   // (cmplx.Real * cmplx.Real) + (cmplx.Imaginary * cmplx.Imaginary);
var angle = Atan(cmplx);                    // System.Math.Atan2(cmplx.Imaginary, cmplx.Real);
```

Reply:
`amplitude` can be replaced by the `cmplx.Amplitude` property and `angle` can be replaced by `Amplitude.Phase`.

Reply:
FYI (I have just noticed your last note in this discussion):
```c#
var amplitude = cmplx.Amplitude;
```
Compiler error:
'Complex' does not contain a definition for 'Amplitude' and no accessible extension method 'Amplitude' accepting a first argument of type 'Complex' could be found (are you missing a using directive or an assembly reference?) [Microsoft.Quantum.Simulators]csharp(CS1061)

I also fail to find "Amplitude" in [Complex](https://docs.microsoft.com/en-us/dotnet/api/system.numerics.complex?view=netcore-3.1), its [Properties](https://docs.microsoft.com/en-us/dotnet/api/system.numerics.complex?view=netcore-3.1#properties), etc.

Reply:
Sorry, I should have written `Magnitude` instead of `Amplitude`.

Reply:
Ah, now I understand. Thank you.
From the very beginning, prior to my change, the [line](https://github.com/microsoft/qsharp-runtime/blob/5ea517ccd953bfea1609b6ebe176bde66576830c/src/Simulation/Simulators/QuantumSimulator/StateDumper.cs#L149) 
```c#
var amplitude = (real * real) + (img * img);
```
should have been
```c#
var magnitude = Math.Sqrt(real * real + img * img); // `Math.Sqrt()` was missing, and named magitude.
```
I'll simplify during stage 3.